internal/getproviders/package_location_http_archive.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package getproviders
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/hashicorp/go-getter"
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/opentofu/opentofu/internal/httpclient"
+	"github.com/opentofu/opentofu/internal/logging"
+)
+
+// PackageHTTPURL is a provider package location accessible via HTTP.
+//
+// Its value is a URL string using either the http: scheme or the https: scheme.
+// The URL should respond with a .zip archive whose contents are to be extracted
+// into a local package directory.
+type PackageHTTPURL string
+
+var _ PackageLocation = PackageHTTPURL("")
+
+func (p PackageHTTPURL) String() string { return string(p) }
+
+func (p PackageHTTPURL) InstallProviderPackage(ctx context.Context, meta PackageMeta, targetDir string, allowedHashes []Hash) (*PackageAuthenticationResult, error) {
+	url := meta.Location.String()
+
+	// When we're installing from an HTTP URL we expect the URL to refer to
+	// a zip file. We'll fetch that into a temporary file here and then
+	// delegate to installFromLocalArchive below to actually extract it.
+	// (We're not using go-getter here because its HTTP getter has a bunch
+	// of extraneous functionality we don't need or want, like indirection
+	// through X-Terraform-Get header, attempting partial fetches for
+	// files that already exist, etc.)
+
+	retryableClient := retryablehttp.NewClient()
+	retryableClient.HTTPClient = httpclient.New()
+	retryableClient.RetryMax = maxHTTPPackageRetryCount
+	retryableClient.RequestLogHook = func(logger retryablehttp.Logger, _ *http.Request, i int) {
+		if i > 0 {
+			logger.Printf("[INFO] failed to fetch provider package; retrying")
```

Comment:
food for thought: Do you think it would be nice to include the number of retries here?

Maybe something like 
```go
logger.Printf("[INFO] failed to fetch provider package; retrying attempt %s/%s", i, maxHTTPPackageRetryCount)
```